### PR TITLE
Add a specification for Composition Functions

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes

Relates to https://github.com/crossplane/crossplane/issues/3714

This is a first pass at a 'formal' spec for a Composition Function. I think it will be important for all Functions to behave in a similar way in order for us to build tooling around them.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Added or updated unit **and** E2E tests for my change.
- [ ] ~Run `make reviewable` to ensure this PR is ready for review.~
- [ ] ~Added `backport release-x.y` labels to auto-backport this PR, if necessary.~
- [ ] ~Opened a PR updating the [docs], if necessary.~

[contribution process]: https://git.io/fj2m9
[docs]: https://docs.crossplane.io/contribute/contribute

### CODE REVIEW

**Comment by negz** [comment]
@plumbis Added you as optional to review this for readability.

File: contributing/specifications/functions.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+Function pipeline, including preventing any Functions that would normally run
+after it from running.
+
+A Function MUST NOT return a Warning or Normal result every time it is called. A
+Function SHOULD return Warning or Normal results to indicate transitions - e.g.
+when entering a warning or normal state.
+
+## Configuration
+
+A Function MUST support the following command-line flags:
```
Comment (phisco) [inline contributing/specifications/functions.md:71]: Maybe we could add something about debug logging, this way we could even think of making that configurable at Function level.

File: contributing/specifications/functions.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+A Function MUST have a name beginning with `function-`. This name MUST appear in
+its package metadata, and SHOULD be used elsewhere, for example its GitHub
+repository.
+
+## Runtime Environment
+
+A Function MUST NOT assume it is deployed in any particular way, for example
+that it is running as a Kubernetes Pod in the same cluster as Crossplane.
+
+A Function SHOULD fail gracefully if it does not have network access, for
```
Comment (turkenh) [inline contributing/specifications/functions.md:103]:
```suggestion
A Function SHOULD fail gracefully if it needs but does not have network access, for
```

I assume this is not relevant for an offline function.

File: contributing/specifications/functions.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+desired state passed to it by previous Functions.
+
+A Function MUST NOT specify desired composite resource spec or metadata fields.
+A Function MAY only specify desired composite resource status fields.
+
+A Function MUST NOT specify desired composed resource status fields. A Function
+MAY only specify non-status desired composed resource fields, for example fields
+under the top-level metadata, spec, or data fields.
+
+A Function SHOULD NOT specify a metadata.name for desired composed resources.
```
Comment (plumbis) [inline contributing/specifications/functions.md:51]: SHOULD NOT or MUST NOT? This reads like a `MUST NOT`

**Review by plumbis — APPROVED** [review]

File: contributing/specifications/functions.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
+# Composition Functions Specification
+
+A Composition Function is a kind of Crossplane extension. A Function is called
+to instruct Crossplane how a Composite Resource (XR) should be composed of other
+resources. This document specifies how a Function MUST behave. Refer to the
+Composition Functions [design document] for broader context on the feature.
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are
+to be interpreted as described in [RFC 2119].
```
[the remark this answers is not in the document]
Comment (negz) [reply contributing/specifications/functions.md:10]: I think I should leave this as is. It's copied verbatim from the RFC, which recommends doing just that. FWIW "RFC X" by itself in internet standards does typically mean IETF RFC, so I'm hopeful it won't confuse too many folks.

File: contributing/specifications/functions.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+desired state. For example a Function may:
+
+* Add a new composed resource to the desired state.
+* Add a new field to an existing composed resource in the desired state.
+* Remove a composed resource from the desired state.
+* etc...
+
+A Function MUST do so _intentionally_. Put otherwise, remember that Functions
+are run in a pipeline. A Function MUST NOT accidentally forget to propagate the
+desired state passed to it by previous Functions.
```
[the remark this answers is not in the document]
Comment (negz) [reply contributing/specifications/functions.md:42]: Thanks @plumbis. I might just remove this one - I already say it this way above. What I was going for here was the nuance - it's not that you MUST copy all desired state, but rather that you MUST NOT drop state unintentionally. I agree we should avoid the double negative but not sure how to do so. 🤔

File: contributing/specifications/functions.md (hunk 3 above)
Comment (negz) [reply contributing/specifications/functions.md:51]: This one's actually SHOULD NOT - you can but it's not recommended.

File: contributing/specifications/functions.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+A Function MUST NOT specify desired composite resource spec or metadata fields.
+A Function MAY only specify desired composite resource status fields.
+
+A Function MUST NOT specify desired composed resource status fields. A Function
+MAY only specify non-status desired composed resource fields, for example fields
+under the top-level metadata, spec, or data fields.
```
[the remark this answers is not in the document]
Comment (negz) [reply contributing/specifications/functions.md:49]: Do you think the "top-level" bit is self-explanatory? What I'm going for here is that not everything has metadata and spec - e.g. secrets have metadata and data.

[the remark this answers is not in the document]
Comment (plumbis) [reply contributing/specifications/functions.md:49]: I think it is. If I'm not talking about top-level then it must be under spec, right?

File: contributing/specifications/functions.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeanduplessis) [reply contributing/specifications/functions.md:10]: TIL 👍

File: contributing/specifications/functions.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+  tls.crt), and a CA used to verify clients (ca.crt).
+
+A Function MUST support reading the `--tls-certs-dir` flag from the
+`TLS_SERVER_CERTS_DIR` environment variable. A Function MUST serve insecurely if
+both `--tls-certs-dir` and `--insecure` are specified.
+
+A Function MUST listen for gRPC requests on TCP port 9443, regardless of whether
+they are using mTLS transport security or have transport security disabled.
+
+A Function MUST not disable gRPC transport security unless the `--insecure` flag
```
Comment (bobh66) [inline contributing/specifications/functions.md:84]: MUST NOT

File: contributing/specifications/functions.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (negz) [reply contributing/specifications/functions.md:49]: @plumbis Not necessarily - e.g. a secret has no `spec`, only `data`. The spec and status thing is more of a very prevalent pattern, not a requirement.

File: contributing/specifications/functions.md (hunk 3 above)
Comment (bobh66) [reply contributing/specifications/functions.md:51]: Is it worth adding something like "if a `metadata.name` is specified it MUST be unique for the associated GVK within the kubernetes cluster"?   I have had several questions in Slack this week from people who want to set their own `metadata.name` values for XRs and MRs.

**Review by bobh66 — APPROVED** [review]
LGTM - now I just need to get started on that Python SDK

**Pull request merged**